AWS: Add support for Vivid Vervet & Debian Jessie

justinsb opened the pull request:
Particularly interesting because this is the first non-Redhat systemd system, so we need to fix a few systemd/salt problems and we also can't assume that redhat <=> systemd.

This did let me test btrfs.  _Still_ doesn't support devicemapper though (we need a custom Docker build, it seems) :-(

This depends on the wheezy support, so will need a rebase when that merges.  Please ignore the wheezy commit / stuff!

justinsb committed:
AWS: Add jessie support to cluster script

roberthbailey commented:
Generally looks good. Please ping when you rebase for final review.

roberthbailey commented:
Poke -- needs a rebase.

justinsb commented:
@roberthbailey - this is dependent on 9371 anyway.  Should I rebase before that merges?

roberthbailey commented:
Probably not. I just hadn't seen any movement in a few days and thought this might be ready for review (trying to clear out my queue).

justinsb commented:
@roberthbailey #9371 merged, so I rebased and hopefully we can get it out of your queue now!

roberthbailey commented:
LGTM.

fabioy commented:
@k8s-bot ok to test

justinsb commented:
I'm guessing this is because of my Salt change, but I need to figure it out.  I think this might be fairly important because it ties in to #9675, in particular if we want to run an OS that includes a more modern kernel on AWS, which typically implies systemd.

Hoping I will get to this today or tomorrow.

justinsb commented:
I think the problem is #9837 (although that shouldn't be affecting GCE, but I can't exactly figure out my Salt change while Debian is now broken).  Going to mark this as WIP for now.  I'm going to look at targeting the latest Ubuntu, which has a newer kernel and uses systemd.

My goals is to find a maintained image on AWS that has a sufficiently modern kernel that works without a reboot (I've tried doing a reboot as part of bringup; it is OK but not great).

justinsb committed:
AWS: Initial support for Ubuntu Vivid Vervet

justinsb committed:
AWS: Ubuntu Vivid, set cgroup_root to docker

Otherwise libcontainer generates systemd unit file names that start with "-",
which are not valid.

justinsb committed:
AWS: Change default OS to Ubuntu Vivid Vervet

The background for this change is in #9675.

In short, Vivid Vervet gives us a supported/updated image,
that runs Docker with a working storage engine, but doesn't
require a reboot as part of node start.

Fixes #9675.

justinsb commented:
Unblocked this by taking systemd service management out of Salt, into our own script, in the face of apparently insurmountable Salt bugs (#9675): https://github.com/justinsb/kubernetes/commit/5bcbcecc604fcd33d6913f7279e0211a85091f47

Also added support for Ubuntu Vivid Vervet on AWS, which seems to be the only AWS AMI which both had a modern enough kernel to run Docker and didn't require a reboot (#9892)

I'm going to remove the lgtm tag, because the linked commit above is non-trivial, and I don't want to "sneak this one in".  I think the changes to salt are systemd only, so will only affect AWS & vagrant:  cc @derekwaynecarr

derekwaynecarr commented:
Please hold this until tomorrow when I have a chance to review and test.

Sent from my iPhone

> On Jun 16, 2015, at 6:41 PM, Justin Santa Barbara <notifications@github.com> wrote:
> 
> Unblocked this by taking systemd service management out of Salt, into our own script, in the face of apparently insurmountable Salt bugs (#9675): justinsb@5bcbcec
> 
> Also added support for Ubuntu Vivid Vervet on AWS, which seems to be the only AWS AMI which both had a modern enough kernel to run Docker and didn't require a reboot (#9892)
> 
> I'm going to remove the lgtm tag, because the linked commit above is non-trivial, and I don't want to "sneak this one in". I think the changes to salt are systemd only, so will only affect AWS & vagrant: cc @derekwaynecarr
> 
> —
> Reply to this email directly or view it on GitHub.
>

justinsb commented:
Grr... I thought I had tested this on a _non_-systemd system, but I obviously hadn't.  Or GCE's wheezy is odd in some way.  I debugged and (on GCE) the systemd pillar variables appeared to be set; I'm trying out using a grain which is apparently more salty and will definitely be easier to debug in #9935.

justinsb commented:
Fixed the problem - I made a stupid mistake: https://github.com/justinsb/kubernetes/commit/0feb0687bf86a0082361665a173dfe695f6b2c87

(and this also explains the testing mystery; I obviously hadn't tested where docker_opts was not set).

@zmerlynn suggests that grains is indeed more salty (https://github.com/GoogleCloudPlatform/kubernetes/pull/9935#issuecomment-112811947).

So I'm going to force-push the grainy version into this PR.

For reference, the SHA with pillar was f7bd060492bc0ef2408d6b9df71ff9a9c0898167

justinsb commented:
The test status here is a bit confusing.  34a (passed on GCE) is the latest.  Shippable has failed with a flake, I believe.  The most recent GCE failure as on an earlier commit, but it took longer to run than the passing test run which started after it. 

In short, this passes GCE & I think the Shippable fail is a flake (but I can't see it to be sure)

justinsb commented:
Rebased to double-check that it does pass testing (it does!)

roberthbailey commented:
@derekwaynecarr Have you had a chance to take a look?

derekwaynecarr commented on cluster/saltbase/salt/base.sls:
```diff
@@ -10,4 +10,24 @@ pkg-core:
\ No newline at end of file
+{% endif %}
+
+{% if grains['oscodename'] in [ 'vivid', 'jessie' ] %}
```
I like this.

derekwaynecarr commented on cluster/saltbase/salt/docker/docker-defaults:
```diff
@@ -1,6 +1,3 @@
-DOCKER_OPTS="${DOCKER_OPTS} {{grains.docker_opts}}"
-{% endif %}
-DOCKER_OPTS="${DOCKER_OPTS} --bridge cbr0 --iptables=false --ip-masq=false"
+{% set grains_opts = grains.get('docker_opts', '') -%}
```
Is this a way of giving a default value in salt?

derekwaynecarr commented:
The salt changes look good, I like the systemd abstraction.  If this can hold to Monday, I can test spinning up a cluster that morning. 

Sent from my iPhone

> On Jun 19, 2015, at 7:28 PM, Robert Bailey <notifications@github.com> wrote:
> 
> @derekwaynecarr Have you had a chance to take a look?
> 
> —
> Reply to this email directly or view it on GitHub.
>

roberthbailey commented:
I'm not going to merge it on a friday night. Let me know how the testing goes monday.

derekwaynecarr commented:
I tried this out this morning, and it appears that the Kubelet is not configuring properly for me.  Right now, it appears to try and put the stuff for the Kubelet in ```/etc/init.d/*``` so I need to review this more to see why things are messed up for me.

derekwaynecarr commented:
I am seeing the following behavior.

On an initial ```kube-up```, we are writing the ```is_systemd``` grain properly, but the grains that are written are not reloaded or used unless I do a subsequent ```kube-push```.  I think if we are dynamically writing grains during provisioning process, we need to sync grains, and according to this document:

http://docs.saltstack.com/en/latest/topics/targeting/grains.html

It only happens when you call sync to high state which explains the behavior I am seeing.

I think this may work more consistently if we output the grain information outside of salt, and instead just add a new grain to our grains.conf files as part of initial provisioning.  In this way, ```kube-up``` will work on the initial provision of the cluster.

justinsb commented:
@derekwaynecarr sorry about the problem.  How did you hit this - was it simply by trying to kube-up with vagrant?  Let me know and I can make sure it is fixed.

In any case, I moved from grains to pillars (back to pillars, actually).  The problem with pre-setting it in grains is that all the scripts would have to set it correctly.  I tried adding (redhat or is_systemd) to maximize compatibility, but there would still have been a problem on the systemd_system_path.

If you're able to test it out or let me know how you hit a problem before, that would be great appreciated!

justinsb committed:
AWS: Treat ubuntu as an alias for 'latest ubuntu', i.e. 'vivid'

derekwaynecarr commented:
I rebased your PR and then did kube-up with vagrant provider.  It fails because kubelet is misconfigured, and then a kube-push things get better (the whole systemd stuff works again).

Sent from my iPhone

> On Jun 24, 2015, at 10:59 PM, Justin Santa Barbara <notifications@github.com> wrote:
> 
> @derekwaynecarr sorry about the problem.  How did you hit this - was it simply by trying to kube-up with vagrant? Let me know and I can make sure it is fixed.
> 
> In any case, I moved from grains to pillars (back to pillars, actually). The problem with pre-setting it in grains is that all the scripts would have to set it correctly. I tried adding (redhat or is_systemd) to maximize compatibility, but there would still have been a problem on the systemd_system_path.
> 
> If you're able to test it out or let me know how you hit a problem before, that would be great appreciated!
> 
> —
> Reply to this email directly or view it on GitHub.
>

justinsb commented:
@derekwaynecarr I take you're referring to the situation when I used grains.  I pushed an updated version using pillar, and the cluster comes up now on vagrant (with #10347, but that is a separate issue I think)

justinsb commented:
@k8s-bot ok to test

justinsb commented:
Persuaded shippable to pass, but I don't think we have a recent e2e run.  Anyone know how to persuade k8sbot to test it?

roberthbailey commented:
@k8s-bot Test this please.

roberthbailey commented:
@justinsb this needs a rebase.

justinsb commented:
@k8s-bot retest this please

justinsb commented:
If it would be helpful, I can split this into two PRs - the (higher risk) systemd fixes and the (more contained) AWS changes.  The AWS work relies on the systemd salt fixes, but splitting would make it easier to focus on the high-risk work.

derekwaynecarr commented:
@justinsb i think splitting the systemd parts out would help.

justinsb commented:
I split the systemd work out into #10488

justinsb commented:
Rebasing this one now

justinsb commented:
Rebased.  This is now AWS-only, except for the the change to `cluster/saltbase/salt/kubelet/default`, which is gated on Ubuntu Vivid Vervet (which I don't think anyone but AWS uses).

benmccann commented:
I'm curious why both Jessie and Vivid? Would it be easier to support only Vivid or are there people that have asked for Jessie?

justinsb commented:
@benmccann I originally thought Jessie was the safer bet, but it turned out to require a reboot (to enable the memory controller, as I recall).  Vivid was the best image I could find: no reboot needed, modern kernel that supports docker & good filesystem choices.

I don't know whether it is worth dropping Jessie support; I'm hoping they default the memory controller on in another release.  If we drop Jessie we should probably also drop trusty & wheezy; I'd rather keep them in there for now and see how people use it!

justinsb commented:
Weird error in GCE build log: `Unable to query GitHub for status of PullRequestjava.io.IOException: Server returned HTTP response code: 401 for URL: https://api.github.com/repos/GoogleCloudPlatform/kubernetes/pulls/9381`

@k8s-bot test this please

benmccann commented:
@justinsb cool. i just didn't want people getting confused by using a non-optimal config, but i missed that it's setting the default to vivid and you have good docs around it, so i think it's great the way you have it setup now

roberthbailey commented:
Bouncing to @brendandburns since he's been looking at AWS lately.

justinsb commented:
@brendanburns: to bring you up to speed: this is my last "feature" PR for v1.  (There may be bugfix PRs, but I don't know of any, so fingers crossed!)  It switches the default OS from Ubuntu 14.04 to Ubuntu 15.04, which is the OS I've been testing primarily with AWS recently.  It is the only image that I have found that includes a sufficiently modern kernel to keep Docker happy, that doesn't require a reboot (which rules out Debians).  It also supports different filesystems, including btrfs which I have found to be more reliable than aufs on more recent kernels.  The systemd work we merged earlier today was all to support this, because 15.04 uses systemd.

With this, I have one known-failing e2e test (`Services should be able to change the type and nodeport settings of a service`), and some flakes/semi-random failures.  That failing test is one I wrote myself to cover a known-problem with AWS recreating the ELB whenever you change a port, fixed in PR #8366 but (rightly) deferred to post-v1.  (GCE does not have the problem because it reuses the same IP)

I think it's a good PR!  I've been testing the end-result pretty heavily, and it solves a lot of real problems.  Also low-risk: almost entirely in cluster/aws, one small change in salt that is conditional on detecting 15.04.

brendandburns commented on cluster/aws/jessie/util.sh:
```diff
@@ -0,0 +1,81 @@
+#
+# Vars set:
+#   AWS_IMAGE
+function detect-jessie-image () {
```
Can we add a comment about how insane it is that AWS forces you to do this :P

justinsb replied:
With pleasure!  How much vitriol may I use?

brendandburns commented:
LGTM.  ok-to-merge
